### ISSUE TITLE
Investigate if we can use gradient clipping when optimizer in backward in True

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#1934]
Currently our optimizer in backward is not compatible with gradient clipping. This may be be an API/implementation issue, and not a technical limitation.

from full_finetune_distributed
```
# For optimizer in backward, we need to normalize before calling backward
# This case and gradient accumulation are mutually exclusive
if self._optimizer_in_bwd:
    torch.distributed.all_reduce(num_tokens)
    torch.distributed.all_reduce(running_loss)
    current_loss = current_loss / num_tokens

current_loss.backward()

# Step with optimizer
if (idx + 1) % self._gradient_accumulation_steps == 0:
    if not self._optimizer_in_bwd:
        # Get total number of tokens across all ranks to normalize gradients
        torch.distributed.all_reduce(num_tokens)
        # This will ensure that the logged loss matches what we're optimizing
        torch.distributed.all_reduce(running_loss)
        # Manually scale the gradients from unnormalized loss by total # of tokens
        training.scale_grads(self._model, 1 / num_tokens)
    if self._clip_grad_norm is not None:
        if self._optimizer_in_bwd:
            raise NotImplementedError(
                "Gradient clipping is not supported after optimizer-in-the-backward."
            )
        grad_norm = torch.nn.utils.clip_grad_norm_(
            self._model.parameters(),
            max_norm=float(self._clip_grad_norm),
        )
    if not self._optimizer_in_bwd:
        self._optimizer.step()
        self._optimizer.zero_grad(set_to_none=True)
```

**Comment by calvinpelletier** [comment]
I believe we just need to add `torch.nn.utils.clip_grad_norm_([param], max_norm=clip_grad_norm)` before this line: https://github.com/pytorch/torchtune/blob/main/torchtune/training/memory.py#L237

**Comment by felipemello1** [comment]
it makes sense to me :)

**Comment by gau-nernst** [comment]
This is a technical limitation. To do gradient clipping, we need to calculate L2 norm of all gradients together before optimizer step. In other words, we must keep all gradients exist in memory. Optimizer in backward will clear gradients progressively, hence gradient clipping is not possible (or to be more exact, the "effective" gradient can only be computed once we have materialized all the gradients -> optimizer in backward is not possible).

Adding `clip_grad_norm()` inside optim in backward will not work (or at least will not work as expected), because it is clipping per-parameter gradient norm, not model gradient norm.

**Comment by ebsmothers** [comment]
Yeah +1 to @gau-nernst's comment. There's no way to calculate the grad norm correctly, I think we can close this issue
